### PR TITLE
INT-3952: TCP: Add Local IP Address Header

### PR DESCRIPTION
**Pull request by garyrussell**
JIRA: https://jira.spring.io/browse/INT-3952

**cherry-pick the first commit to 4.2.x**

### CODE REVIEW

File: spring-integration-ip/src/main/java/org/springframework/integration/ip/tcp/connection/TcpConnection.java [hunk 1]
```diff
@@ -125,4 +125,12 @@
 	 */
 	SSLSession getSslSession();
 
+	/**
+	 * Provides getters for {@link Socket} properties.
+	 * @return the socketInfo - may be null, for example in interceptors; interceptors
+	 * should override and delegate to the actual TcpConnection.
+	 * @since 4.3
+	 */
+	public SocketInfo getSocketInfo();
```
Comment (artembilan) [inline spring-integration-ip/src/main/java/org/springframework/integration/ip/tcp/connection/TcpConnection.java]: All members in the interface are `public` by default.
And `static` if it isn't method :smile: 

Will fix on merge.

Minor of, course, but my IDEA warns me on the matter: like code is redundant.

File: spring-integration-ip/src/main/java/org/springframework/integration/ip/tcp/connection/TcpMessageMapper.java [hunk 2]
```diff
@@ -132,6 +132,10 @@ protected final void addStandardHeaders(TcpConnection connection, AbstractIntegr
 			.setHeader(IpHeaders.IP_ADDRESS, connection.getHostAddress())
 			.setHeader(IpHeaders.REMOTE_PORT, connection.getPort())
 			.setHeader(IpHeaders.CONNECTION_ID, connectionId);
+		SocketInfo socketInfo = connection.getSocketInfo();
+		if (socketInfo != null) {
+			messageBuilder.setHeader(IpHeaders.LOCAL_ADDRESS, socketInfo.getLocalAddress());
```
Comment (artembilan) [inline spring-integration-ip/src/main/java/org/springframework/integration/ip/tcp/connection/TcpMessageMapper.java]: So, have you left `socketInfo` header population as an option for the custom `TcpMessageMapper.supplyCustomHeaders()` implementation ?

**Comment by artembilan** [comment]
That's all merging with that simple fix for the interface.
And cherry-picking, of course, the first commit.

You are still responsible to answer to my question :smile:

**Comment by artembilan** [comment]
Merged as https://github.com/spring-projects/spring-integration/commit/91387147c94edeb1736143d512a89c8fca9679d7
and cherry-picked as https://github.com/spring-projects/spring-integration/commit/1b93bf934e0ac4dcfad59938be30bb836e46b8eb

File: spring-integration-ip/src/main/java/org/springframework/integration/ip/tcp/connection/TcpMessageMapper.java (hunk 2 above)
Comment (garyrussell) [inline spring-integration-ip/src/main/java/org/springframework/integration/ip/tcp/connection/TcpMessageMapper.java]: Yes; I'd rather not carry a live reference to the socket in messages by default - the whole thing will be dropped if serialized anyway. In custom mappers, I expect users to only populate the results of the getters on `SocketInfo` - not the whole thing.

Comment (artembilan) [inline spring-integration-ip/src/main/java/org/springframework/integration/ip/tcp/connection/TcpMessageMapper.java]: :+1:

**Pull request closed without merging**